### PR TITLE
Integrate ccm-nutanix to OCP via CCCMO

### PR DESCRIPTION
**Pull request by yanhua121**
Integrate ccm-nutanix to OCP via CCCMO

### CODE REVIEW

File: manifests/image-references [hunk 1]
```diff
@@ -42,3 +42,7 @@ spec:
     from:
       kind: DockerImage
       name: quay.io/openshift/origin-powervs-cloud-controller-manager
+  - name: nutanix-cloud-controller-manager
+    from:
+      kind: DockerImage
+      name: quay.io/openshift/origin-nutanix-cloud-controller-manager
```
Comment (JoelSpeed) [inline manifests/image-references:48]: This must not be merged until the image is in payload, you may want to split this into a separate PR to get the other parts of this PR merged earlier.

File: pkg/cloud/cloud_test.go [hunk 2]
```diff
@@ -53,6 +53,7 @@ func (tp *testPlatform) getOperatorConfig() config.OperatorConfig {
 			CloudControllerManagerOpenStack: "registry.ci.openshift.org/openshift:openstack-cloud-controller-manager",
 			CloudControllerManagerVSphere:   "registry.ci.openshift.org/openshift:vsphere-cloud-controller-manager",
 			CloudControllerManagerPowerVS:   "quay.io/openshift/origin-powervs-cloud-controller-manager",
+			CloudControllerManagerNutanix:   "registry.ci.openshift.org/openshift:nutanix-cloud-controller-manager",
```
Comment (JoelSpeed) [inline pkg/cloud/cloud_test.go:56]: I think this should be `quay.io/openshift/origin-nutanix-cloud-controller-manager` as is the format with Azure, not sure why these are mixed 🤔

**Review by JoelSpeed — COMMENTED** [review]
Most of this looks ok so we can probably split the image reference out and get the rest merged

File: pkg/cloud/cloud_test.go (hunk 2 above)
Comment (yanhua121) [reply pkg/cloud/cloud_test.go]: Changed to quay.io/openshift/origin-nutanix-cloud-controller-manager

**Comment by JoelSpeed** [comment]
This looks ok but cannot merge with the image-references in there until the image is in payload
/hold

**Comment by elmiko** [comment]
this PR should enable the image in the payload, https://github.com/openshift/ocp-build-data/pull/2276

**Comment by elmiko** [comment]
we have good confidence that the image is being built, i am removing the hold
/hold cancel

**Review by elmiko — COMMENTED** [review]
this is looking good to me, but i am curious why some of the azure files are being updated?

**Comment by lobziik** [comment]
Commit history doesn't look right to me, can you please rebase nutanix related changes on top of the actual master branch?

File: manifests/image-references (hunk 1 above)
Comment (yanhua121) [reply manifests/image-references:48]: Now the nutanix-ccm image is in payload.

**Comment by lobziik** [comment]
this is looking good to me, but I would like to enable nutanix e2e for this repo and check the results before merging this.
WDYT folks @elmiko @JoelSpeed ?

**Comment by lobziik** [comment]
> Do we have working Nutanix E2E yet? Perhaps a pre-merge acceptance from QE would be sufficient for now?

Seems yes, I found some passed job for MAPI, but it's a month and half old though: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_machine-api-operator/1076/pull-ci-openshift-machine-api-operator-master-e2e-nutanix/1580529957895933952

**Comment by JoelSpeed** [comment]
If the CI is working, setting it up as a manual run job here sounds reasonable

I'm not sure who the QE contact would be for this, do we know if there's someone specific for Nutanix? Cc @sunzhaohua2

**Comment by elmiko** [comment]
> this is looking good to me, but I would like to enable nutanix e2e for this repo and check the results before merging this. WDYT folks @elmiko @JoelSpeed ?

i think it's a good idea, we did have some dns (i think) issues with the nutanix ci cluster in the past and i'm not sure if those have been fixed yet. ideally they won't affect the ccm tests.

**Comment by sunzhaohua2** [comment]
> I'm not sure who the QE contact would be for this, do we know if there's someone specific for Nutanix?

@huali9 Could you help to test?

**Comment by huali9** [comment]
> > I'm not sure who the QE contact would be for this, do we know if there's someone specific for Nutanix?
> 
> @huali9 Could you help to test?

Sure, I'll check this.

**Comment by huali9** [comment]
I cannot build image with this pr using cluster-bot. 
job [build openshift/cluster-cloud-controller-manager-operator#201](https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-aws-modern/1597760251623378944) failed
 get error:
`error: unable to create a release: operator "cluster-cloud-controller-manager-operator" contained an invalid image-references file: no input image tag named "nutanix-cloud-controller-manager"`
Seems the cluster-bot still build image from 4.12, I just submitted a [pr](https://github.com/openshift/ci-chat-bot/pull/326) to change it to 4.13, will build image again when the cluster-bot pr get merged. Thanks!

**Comment by huali9** [comment]
@JoelSpeed I cannot build image with this pr using cluster-bot. Can you please help to take a look? thanks!
job [build 4.13,openshift/cluster-cloud-controller-manager-operator#201](https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-aws-modern/1597804670330343424) failed
get error:
`error: unable to create a release: operator "cluster-cloud-controller-manager-operator" contained an invalid image-references file: no input image tag named "nutanix-cloud-controller-manager"`

**Comment by JoelSpeed** [comment]
I don't really know why that's failed, I suspect because there's no image reference currently merged, when you build a payload from the existing payload, the image isn't there. We may need to include the image reference before we can build an image from this via cluster bot

I'll see if we can get @lobziik's CI build together to try and test from there

**Comment by JoelSpeed** [comment]
Looking at the help text, we should probably be able to test this using a CI build
```
launch 4.13.0-0.ci,openshift/cluster-cloud-controller-manager-operator#201
```

**Comment by elmiko** [comment]
this is really odd because i can see the image being created in our ci cluster, it appears as an image stream tag in the ART 4.13 builds. should we reach out to them?

**Comment by JoelSpeed** [comment]
We should probably reach out yes, though I suspect because nothing is referencing it, the image is being pruned from the manifest list as an optimisation, which makes this a little tricky to test

**Comment by elmiko** [comment]
@sosiouxme we had talked about this image on slack, do you have any guidance here?

**Comment by JoelSpeed** [comment]
/hold

There appears to be something broken about the cloud config sync when we ran the nutanix e2e job, please investigate

**Comment by yanhua121** [comment]
> @yanhua121 Just to be clear, this isn't a test flake, there is some investigation needed here. Please look into why the CCMO is reporting the cluster config controller as degraded when running on Nutanix. Perhaps you missed a switch for the platform within the config-sync-controller or perhaps that controller isn't actually needed to operate on Nutanix. Please check the logs to diagnose further

@JoelSpeed It seems to pass the e2e-nutanix test run, we need to first merge https://github.com/openshift/installer/pull/6269, which includes changes to create the Infrastructure.CloudConfig ConfigMap.

**Comment by JoelSpeed** [comment]
> @JoelSpeed It seems to pass the e2e-nutanix test run, we need to first merge https://github.com/openshift/installer/pull/6269, which includes changes to create the Infrastructure.CloudConfig ConfigMap.

Ok great, can you liase with the installer team to try and get that PR merged then, thanks

**Comment by JoelSpeed** [comment]
@yanhua121 Can you please rebase this PR and make sure it's up to date?

**Comment by yanhua121** [comment]
> 

Done rebase.

**Comment by JoelSpeed** [comment]
Looks like there's an issue somewhere, the nutanix CCM cannot find the credentials it needs to operate?
```
I0118 06:44:56.750191       1 manager.go:68] fetching instance metadata for node ci-op-7hbpwzlg-daecd-slgph-worker-mzgj7
E0118 06:44:56.750212       1 node_controller.go:244] Error getting instance metadata for node addresses: secret "nutanix-credentials" not found
```
Is there a credentials request that needs to go inline with this controller?

**Comment by yanhua121** [comment]
>

**Comment by yanhua121** [comment]
> Looks like there's an issue somewhere, the nutanix CCM cannot find the credentials it needs to operate?
> I0118 06:44:56.750191       1 manager.go:68] fetching instance metadata for node ci-op-7hbpwzlg-daecd-slgph-worker-mzgj7
> E0118 06:44:56.750212       1 node_controller.go:244] Error getting instance metadata for node addresses: secret "nutanix-credentials" not found
> Is there a credentials request that needs to go inline with this controller?

@JoelSpeed Nutanix CCM requires the secret "nutanix-credentials" in the namespace "openshift-cloud-controller-manager" to operate. This is similar to the MAPI nutanix-controller requires the "nutanix-credentials" in the namespace "openshift-machine-api". This credentials secret manifest is expected to be manually created when installing the OCP cluster, also similar to that in the namespace "openshift-machine-api". If the e2e-nutanix-ovn failed because of the missing secret "nutanix-credentials" in the namespace "openshift-cloud-controller-manager", I guess we need to add the step to create the secret in the e2e-nutanix-ovn.

**Comment by JoelSpeed** [comment]
Do you know how that was done for Machine API? Can we make sure to update the steps to be the same?

**Comment by yanhua121** [comment]
> Do you know how that was done for Machine API? Can we make sure to update the steps to be the same?

@JoelSpeed Once this PR is merged and the manifests/0000_26_cloud-controller-manager-operator_18_credentialsrequest-nutanix.yaml will be included in the OCP release image. And when installing a Nutanix OCP cluster, users just follow the same steps described at https://access.redhat.com/documentation/en-us/openshift_container_platform/4.11/html/installing/installing-on-nutanix#cco-ccoctl-configuring_preparing-to-install-on-nutanix, and the required nutanix-credentials secrets manifests will be generated from the credentialsrequest yamls extracted from the OCP release image. I guess the e2e-nutanix and e2e-nutanix-ovn are not specifically targeted to test nutanix-ccm. We will add the e2e-nutanix-ccm to test nutanix-ccm.

**Comment by JoelSpeed** [comment]
> I guess the e2e-nutanix and e2e-nutanix-ovn are not specifically targeted to test nutanix-ccm. We will add the e2e-nutanix-ccm to test nutanix-ccm.

They should be able to test the cluster is bootstrapping ok though. So, my concern here is why the credentials request is not being picked up. Somewhere in the steps for creating the cluster, in the release repo config, it must iterate through credentials requests and create the credentials secrets for the cluster right? Do you know anything about how this was set up?

**Comment by JoelSpeed** [comment]
Looks like this needs to be updated? 

https://github.com/openshift/release/blob/master/ci-operator/step-registry/ipi/conf/nutanix/manual-creds/ipi-conf-nutanix-manual-creds-commands.sh

If we can update this script to iterate over nutanix credentialsrequests from the payload, this would make it easier to add new credentials in the future

**Comment by JoelSpeed** [comment]
/lgtm
/hold cancel
/retest required

The nutanix test didn't pass, but i can see now that the CCM is running and isn't reporting errors, so lets get this merged, get the CCM rebased and hopefully that brings us up to date

**Comment by JoelSpeed** [comment]
/override ci/prow/e2e-aws-ovn-upgrade

This change should have no effect on AWS, especially not on upgrades. The failure is related to events from the marketplace operator which is also unrelated, pretty confident this change will not affect the AWS upgrade and that the issue is unrelated
